Mathlib/Data/Set/Basic.lean

```diff
@@ -1521,6 +1521,34 @@ theorem eq_empty_of_ssubset_singleton {s : Set α} {x : α} (hs : s ⊂ {x}) : s
   ssubset_singleton_iff.1 hs
 #align set.eq_empty_of_ssubset_singleton Set.eq_empty_of_ssubset_singleton
 
+theorem Nonempty.eq_singleton_of_unique [Unique α] {s : Set α} (h : s.Nonempty) :
+    s = {default} := h.subset_singleton_iff.mp (fun x _ => by simp [Unique.uniq _ x])
+
+theorem Nonempty.eq_singleton_zero_of_unique[Zero α] [Unique α] {s : Set α} (h : s.Nonempty) :
+    s = {0} := by convert h.eq_singleton_of_unique
+
+theorem Nonempty.eq_singleton_one_of_unique [One α] [Unique α] {s : Set α} (h : s.Nonempty) :
+    s = {1} := by convert h.eq_singleton_of_unique
+
+noncomputable def uniqueOfNonempty [Subsingleton α] {s : Set α} (h : s.Nonempty) :
+    Unique α := uniqueOfSubsingleton h.choose
+
+theorem Nonempty.eq_singleton_of_subsingleton [Subsingleton α] {s : Set α}
+    (h : s.Nonempty) :
+    letI := uniqueOfNonempty h
+    s = {default} := by
+  letI := uniqueOfNonempty h
+  exact h.eq_singleton_of_unique
+
+theorem Nonempty.eq_of_subsingleton [Subsingleton α] {s : Set α} (h : s.Nonempty) (a : α) :
+    s = {a} := by rw [h.eq_singleton_of_subsingleton, Unique.uniq _ a]
+
+theorem Nonempty.eq_zero [Subsingleton α] [Zero α] {s : Set α} (h : s.Nonempty) :
+    s = {0} := h.eq_of_subsingleton 0
+
+theorem Nonempty.subset_eq_one [Subsingleton α] [One α] {s : Set α} (h : s.Nonempty) :
+    s = {1} := h.eq_of_subsingleton 1
+
```

Comment:
I suggest that you only add two lemmas as others are direct applications:
```suggestion
theorem eq_of_nonempty_of_subsingleton {α} [Subsingleton α] (s t : Set α)
    [hs : Nonempty s] [ht : Nonempty t] : s = t := by
  have ⟨a, ha⟩ := hs
  have ⟨b, hb⟩ := ht
  ext; constructor <;> intro
  · exact mem_of_eq_of_mem (Subsingleton.elim _ _) hb
  · exact mem_of_eq_of_mem (Subsingleton.elim _ _) ha

theorem eq_of_nonempty_of_subsingleton' {α} [Subsingleton α] {s : Set α} (t : Set α)
    (hs : s.Nonempty) [Nonempty t] : s = t :=
  have := hs.to_subtype; eq_of_nonempty_of_subsingleton s t

theorem Nonempty.eq_singleton_of_unique [Unique α] {s : Set α} (h : s.Nonempty) :
    s = {default} := eq_of_nonempty_of_subsingleton' {default} h

theorem Nonempty.eq_singleton_zero_of_unique [Zero α] [Unique α] {s : Set α} (h : s.Nonempty) :
    s = {0} := eq_of_nonempty_of_subsingleton' {0} h

theorem Nonempty.eq_singleton_one_of_unique [One α] [Unique α] {s : Set α} (h : s.Nonempty) :
    s = {1} := eq_of_nonempty_of_subsingleton' {1} h

theorem Nonempty.eq_of_subsingleton [Subsingleton α] {s : Set α} (h : s.Nonempty) (a : α) :
    s = {a} := eq_of_nonempty_of_subsingleton' {a} h

theorem Nonempty.eq_zero [Subsingleton α] [Zero α] {s : Set α} (h : s.Nonempty) :
    s = {0} := eq_of_nonempty_of_subsingleton' {0} h

theorem Nonempty.subset_eq_one [Subsingleton α] [One α] {s : Set α} (h : s.Nonempty) :
    s = {1} := eq_of_nonempty_of_subsingleton' {1} h
```

Reply:
LGTM (modulo some small name changes). Thanks!

Reply:
I'm suggesting you only keep the first two lemmas. What do you think? Certainly if you have a [Subsingleton α] lemma you don't need the [Unique α] lemma with otherwise the same assumptions.

Reply:
I like to keep the lemmas `Nonempty.eq_zero` and `Nonempty.eq_zero`. I think it's a very convenient to replace easily any nonempty subset of, say, a trivial ring by {0}. This is actually the reason why I started this PR. I don't care that much about the [Unique] versions.

Reply:
BTW, since you are now listed as co-author, feel free to push any change you want to make ;)

Reply:
Do you use `eq_of_nonempty_of_subsingleton'` as a rewrite lemma? If so, does it not work if you `rw [h.eq_of_nonempty_of_subsingleton' {0}]` instead of `rw [h.eq_zero]`?

Reply:
Probably, but I think it is more natural to use `h.eq_zero`. Especially, since it is a situation that should come up quite often...